test/com/linkedin/parseq/internal/TestPlanCompletionListener.java

```diff
@@ -0,0 +1,121 @@
+package com.linkedin.parseq.internal;
+
+import com.linkedin.parseq.BaseEngineTest;
+import com.linkedin.parseq.EngineBuilder;
+import com.linkedin.parseq.Task;
+import com.linkedin.parseq.trace.ResultType;
+import com.linkedin.parseq.trace.ShallowTrace;
+import com.linkedin.parseq.trace.Trace;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import static com.linkedin.parseq.Task.value;
+
+
+/**
+ * @author Ang Xu
+ */
+public class TestPlanCompletionListener extends BaseEngineTest  {
+
+  private TraceCaptureListener _traceCaptureListener;
+
+  @Override
+  protected void customizeEngine(EngineBuilder engineBuilder) {
+    _traceCaptureListener = new TraceCaptureListener();
+    engineBuilder.setPlanCompletionListener(_traceCaptureListener);
+  }
+
+  @Test
+  public void testSingleTask() {
+    Task<?> task = value("taskName", "value").andThen(value("value2"));
+
+    runAndWait(task);
+
+    Assert.assertTrue(_traceCaptureListener.isDone());
+  }
+
+  @Test
+  public void testUnfinishedTask() {
+    Task<?> task = value("unFinished", "value");
+
+    Assert.assertFalse(_traceCaptureListener.isDone());
+  }
+
+  @Test
+  public void testWithTimerTask() {
+    Task<?> task = delayedValue("value", 5, TimeUnit.SECONDS).withTimeout(2, TimeUnit.SECONDS);
```

Comment:
IIUC this test will take 2s to complete? Can you change timeout value to something much smaller e.g. 50ms?